src/v/datalake/coordinator/coordinator.cc

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/coordinator/coordinator.h"
+
+#include "base/vlog.h"
+#include "container/fragmented_vector.h"
+#include "datalake/coordinator/state_update.h"
+#include "datalake/logger.h"
+#include "model/fundamental.h"
+#include "model/record_batch_reader.h"
+#include "storage/record_batch_builder.h"
+
+namespace datalake::coordinator {
+
+namespace {
+coordinator::errc convert_stm_errc(coordinator_stm::errc e) {
+    switch (e) {
+    case coordinator_stm::errc::not_leader:
+        return coordinator::errc::not_leader;
+    case coordinator_stm::errc::shutting_down:
+        return coordinator::errc::shutting_down;
+    case coordinator_stm::errc::apply_error:
+        return coordinator::errc::stm_apply_error;
+    case coordinator_stm::errc::raft_error:
+        return coordinator::errc::timedout;
+    }
+}
+} // namespace
+
+std::ostream& operator<<(std::ostream& o, coordinator::errc e) {
+    switch (e) {
+    case coordinator::errc::not_leader:
+        return o << "coordinator::errc::not_leader";
+    case coordinator::errc::shutting_down:
+        return o << "coordinator::errc::shutting_down";
+    case coordinator::errc::stm_apply_error:
+        return o << "coordinator::errc::stm_apply_error";
+    case coordinator::errc::timedout:
+        return o << "coordinator::errc::timedout";
+    }
+}
+
+ss::future<> coordinator::stop_and_wait() {
+    as_.request_abort();
+    return gate_.close();
+}
+
+checked<ss::gate::holder, coordinator::errc> coordinator::maybe_gate() {
+    ss::gate::holder h;
```

Comment:
`h` is unused? intentional?

Reply:
Oops! unintentional, may fix in a follow up if this PR can get merged